File: hw/chicaiza/u1/hw14SaveInJson/ChickenFarmSimulatorPython/chickenfarmsimulator/controller/farm_controller.py
import json
import os
from chickenfarmsimulator.model.chicken_farmer import ChickenFarmer
from chickenfarmsimulator.model.chicken_coop import ChickenCoop
from chickenfarmsimulator.model.chicken import Chicken

class FarmController:
    def __init__(self, data_file='data/farms.json'):
        self.data_file = data_file
        self.farmer = None
        self.load_data()
    
    def load_data(self):
        """Carga los datos desde el archivo JSON"""
        if os.path.exists(self.data_file):
            try:
                with open(self.data_file, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    self.farmer = ChickenFarmer.from_dict(data)
            except (json.JSONDecodeError, KeyError):
                # Si hay error en el archivo, crear nuevo farmer
                self.farmer = ChickenFarmer("Default Farmer")
        else:
            self.farmer = ChickenFarmer("Default Farmer")
    
    def save_data(self):
        """Guarda los datos en el archivo JSON"""
        os.makedirs(os.path.dirname(self.data_file), exist_ok=True)
        with open(self.data_file, 'w', encoding='utf-8') as f:
            json.dump(self.farmer.to_dict(), f, indent=2, ensure_ascii=False)
    
    def create_coop(self, coop_id: int):
        """Crea una nueva granja"""
        if self.farmer.get_coop(coop_id):
            return False  # Ya existe una granja con ese ID
        coop = ChickenCoop(coop_id)
        self.farmer.add(coop)
        self.save_data()
        return True
    
    def add_chicken_to_coop(self, coop_id: int, chicken_id: int, name: str, color: str, age: int, is_molting: bool):
        """Añade un pollo a una granja"""
        coop = self.farmer.get_coop(coop_id)
        if not coop:
            return False  # Granja no encontrada
        
        # Verificar si el pollo ya existe
        if coop.get_chicken(chicken_id):
            return False  # Pollo ya existe
        
        chicken = Chicken(chicken_id, name, color, age, is_molting)
        coop.add(chicken)
        self.save_data()
        return True
    
    def get_all_coops(self):
        """Obtiene todas las granjas"""
        return self.farmer.coops
    
    def get_coop(self, coop_id: int):
        """Obtiene una granja específica"""
        return self.farmer.get_coop(coop_id)
    
    def update_chicken(self, coop_id: int, chicken_id: int, name: str, color: str, age: int, is_molting: bool):
        """Actualiza un pollo"""
        coop = self.farmer.get_coop(coop_id)
        if not coop:
            return False
        
        success = coop.update_chicken(chicken_id, name, color, age, is_molting)
        if success:
            self.save_data()
        return success
    
    def delete_chicken(self, coop_id: int, chicken_id: int):
        """Elimina un pollo"""
        coop = self.farmer.get_coop(coop_id)
        if not coop:
            return False
        
        coop.remove(chicken_id)
        self.save_data()
        return True
    
    def perform_chicken_action(self, coop_id: int, chicken_id: int, action: str):
        """Realiza una acción del pollo"""
        coop = self.farmer.get_coop(coop_id)
        if not coop:
            return False
        
        chicken = coop.get_chicken(chicken_id)
        if not chicken:
            return False
        
        actions = {
            'cluck': chicken.cluck,
            'wander': chicken.wander,
            'eat': chicken.eat,
            'drink': chicken.drink,
            'poop': chicken.poop,
            'lay_egg': chicken.lay_an_egg
        }
        
        if action in actions:
            actions[action]()
            return True
        return False
    
    def get_chicken_count(self):
        """Obtiene el número total de pollos"""
        total = 0
        for coop in self.farmer.coops:
            total += len(coop.chickens)
        return total
    
    def get_coop_count(self):
        """Obtiene el número total de granjas"""
        return len(self.farmer.coops)

File: hw/chicaiza/u1/hw14SaveInJson/ChickenFarmSimulatorJavaScript/chicken_farm_simulator/src/controller/FarmController.js
import { ChickenFarmer } from '../model/ChickenFarmer.js';
import { ChickenCoop } from '../model/ChickenCoop.js';
import { Chicken } from '../model/Chicken.js';
import fs from 'fs';
import path from 'path';

export class FarmController {
    constructor(dataFile = 'data/farms.json') {
        this.dataFile = dataFile;
        this.farmer = null;
        this.loadData();
    }
    
    loadData() {
        try {
            if (fs.existsSync(this.dataFile)) {
                const data = JSON.parse(fs.readFileSync(this.dataFile, 'utf8'));
                this.farmer = ChickenFarmer.fromDict(data);
            } else {
                this.farmer = new ChickenFarmer("Default Farmer");
            }
        } catch (error) {
            console.log("Error loading data, creating new farmer...");
            this.farmer = new ChickenFarmer("Default Farmer");
        }
    }
    
    saveData() {
        try {
            const dir = path.dirname(this.dataFile);
            if (!fs.existsSync(dir)) {
                fs.mkdirSync(dir, { recursive: true });
            }
            fs.writeFileSync(this.dataFile, JSON.stringify(this.farmer.toDict(), null, 2));
            return true;
        } catch (error) {
            console.log("Error saving data:", error.message);
            return false;
        }
    }
    
    createCoop(coopId) {
        if (this.farmer.getCoop(coopId)) {
            return false;
        }
        const coop = new ChickenCoop(coopId);
        this.farmer.add(coop);
        return this.saveData();
    }
    
    addChickenToCoop(coopId, chickenId, name, color, age, isMolting) {
        const coop = this.farmer.getCoop(coopId);
        if (!coop) {
            return false;
        }
        
        if (coop.getChicken(chickenId)) {
            return false;
        }
        
        const chicken = new Chicken(chickenId, name, color, age, isMolting);
        coop.add(chicken);
        return this.saveData();
    }
    
    getAllCoops() {
        return this.farmer.coops;
    }
    
    getCoop(coopId) {
        return this.farmer.getCoop(coopId);
    }
    
    updateChicken(coopId, chickenId, name, color, age, isMolting) {
        const coop = this.farmer.getCoop(coopId);
        if (!coop) {
            return false;
        }
        
        const success = coop.updateChicken(chickenId, name, color, age, isMolting);
        if (success) {
            return this.saveData();
        }
        return false;
    }
    
    deleteChicken(coopId, chickenId) {
        const coop = this.farmer.getCoop(coopId);
        if (!coop) {
            return false;
        }
        
        coop.remove(chickenId);
        return this.saveData();
    }
    
    performChickenAction(coopId, chickenId, action) {
        const coop = this.farmer.getCoop(coopId);
        if (!coop) {
            return false;
        }
        
        const chicken = coop.getChicken(chickenId);
        if (!chicken) {
            return false;
        }
        
        const actions = {
            'cluck': () => chicken.cluck(),
            'wander': () => chicken.wander(),
            'eat': () => chicken.eat(),
            'drink': () => chicken.drink(),
            'poop': () => chicken.poop(),
            'lay_egg': () => chicken.layAnEgg()
        };
        
        if (actions[action]) {
            actions[action]();
            return true;
        }
        return false;
    }
    
    getChickenCount() {
        return this.farmer.coops.reduce((total, coop) => total + coop.chickens.length, 0);
    }
    
    getCoopCount() {
        return this.farmer.coops.length;
    }
}